src/arrow-util/src/builder.rs

```diff
@@ -228,11 +229,178 @@ fn scalar_to_arrow_datatype(scalar_type: &ScalarType) -> Result<(DataType, Strin
+                    false,
+                ),
+                "map",
+            )
+        }
         _ => anyhow::bail!("{:?} unimplemented", scalar_type),
     };
     Ok((data_type, extension_name.to_lowercase()))
 }
 
+fn builder_for_datatype(
+    data_type: &DataType,
+    item_capacity: usize,
+    data_capacity: usize,
+) -> Result<ColBuilder, anyhow::Error> {
+    let builder = match &data_type {
+        DataType::Boolean => {
+            ColBuilder::BooleanBuilder(BooleanBuilder::with_capacity(item_capacity))
+        }
+        DataType::Int16 => ColBuilder::Int16Builder(Int16Builder::with_capacity(item_capacity)),
+        DataType::Int32 => ColBuilder::Int32Builder(Int32Builder::with_capacity(item_capacity)),
+        DataType::Int64 => ColBuilder::Int64Builder(Int64Builder::with_capacity(item_capacity)),
+        DataType::UInt8 => ColBuilder::UInt8Builder(UInt8Builder::with_capacity(item_capacity)),
+        DataType::UInt16 => ColBuilder::UInt16Builder(UInt16Builder::with_capacity(item_capacity)),
+        DataType::UInt32 => ColBuilder::UInt32Builder(UInt32Builder::with_capacity(item_capacity)),
+        DataType::UInt64 => ColBuilder::UInt64Builder(UInt64Builder::with_capacity(item_capacity)),
+        DataType::Float32 => {
+            ColBuilder::Float32Builder(Float32Builder::with_capacity(item_capacity))
+        }
+        DataType::Float64 => {
+            ColBuilder::Float64Builder(Float64Builder::with_capacity(item_capacity))
+        }
+        DataType::Date32 => ColBuilder::Date32Builder(Date32Builder::with_capacity(item_capacity)),
+        DataType::Time64(arrow::datatypes::TimeUnit::Microsecond) => {
+            ColBuilder::Time64MicrosecondBuilder(Time64MicrosecondBuilder::with_capacity(
+                item_capacity,
+            ))
+        }
+        DataType::Timestamp(arrow::datatypes::TimeUnit::Microsecond, _) => {
+            ColBuilder::TimestampMicrosecondBuilder(TimestampMicrosecondBuilder::with_capacity(
+                item_capacity,
+            ))
+        }
+        DataType::LargeBinary => ColBuilder::LargeBinaryBuilder(LargeBinaryBuilder::with_capacity(
+            item_capacity,
+            data_capacity,
+        )),
+        DataType::FixedSizeBinary(byte_width) => ColBuilder::FixedSizeBinaryBuilder(
+            FixedSizeBinaryBuilder::with_capacity(item_capacity, *byte_width),
+        ),
+        DataType::Utf8 => {
+            ColBuilder::StringBuilder(StringBuilder::with_capacity(item_capacity, data_capacity))
+        }
+        DataType::LargeUtf8 => ColBuilder::LargeStringBuilder(LargeStringBuilder::with_capacity(
+            item_capacity,
+            data_capacity,
+        )),
+        DataType::Decimal128(precision, scale) => ColBuilder::Decimal128Builder(
+            Decimal128Builder::with_capacity(item_capacity)
+                .with_precision_and_scale(*precision, *scale)?,
+        ),
```

Comment:
these lines were simply moved from the `ArrowColumn::new` method below -- the handling of `DataType::[List, Struct, Map]` right below this is new